io-sim/src/Control/Monad/IOSim.hs

```diff
@@ -179,6 +182,18 @@ instance Monad (SimM s) where
 instance MonadFail (SimM s) where
   fail msg = SimM $ \_ -> Throw (toException (IO.Error.userError msg))
 
+instance MonadFix (SimM s) where
+  mfix act = do
+    -- Impl much like that of System.IO.Unsafe.unsafeFixIO.
+    -- Though based on the comments in fixST if we want to be safe with
+    -- parallel spark evaluation then we might need to use MVars like fixST.
+    (ref, ans) <- liftST $ do
+      ref <- StrictST.newSTRef (throw NonTermination)
+      ans <- StrictST.unsafeInterleaveST (StrictST.readSTRef ref)
+      return (ref, ans)
+    result <- act ans
+    liftST $ StrictST.writeSTRef ref result
+    return result
```

Comment:
My first go was to use a `Fix` constructor, but this implementation is actually much simpler. The impl in the simulator itself cannot be any simpler as far as I can see. It cannot just use `mfix` in ST. It also has to do the ref "create, defered-read, action, write" pattern. And if you're doing that anyway, then it's simpler to do one layer up re-using the generic LiftST actions rather than having special cases of them in the sim interpreter directly.